File: backend/app/schemas/sentiment.py
"""Sentiment schemas for API responses."""

from datetime import datetime

from pydantic import BaseModel

from app.schemas.common import PaginationMeta


class SentimentScoreResponse(BaseModel):
    id: int
    article_id: int
    stock_id: int | None
    label: str
    positive_score: float
    negative_score: float
    neutral_score: float
    model_version: str
    processed_at: datetime
    article_title: str | None = None
    article_source: str | None = None
    article_source_url: str | None = None

    model_config = {"from_attributes": True}


class SentimentSummary(BaseModel):
    """Aggregated sentiment for a ticker or sector."""
    ticker: str | None = None
    sector: str | None = None
    total_articles: int
    positive_count: int
    negative_count: int
    neutral_count: int
    avg_positive: float
    avg_negative: float
    avg_neutral: float
    dominant_label: str


class SentimentTimePoint(BaseModel):
    """Single point in a sentiment time series."""
    date: str
    avg_positive: float
    avg_negative: float
    avg_neutral: float
    article_count: int
    dominant_label: str


class PaginatedSentiment(BaseModel):
    data: list[SentimentScoreResponse]
    meta: PaginationMeta

File: frontend/src/types/sentiment.ts
export interface SentimentScore {
  id: number;
  article_id: number;
  stock_id: number | null;
  label: "positive" | "negative" | "neutral";
  positive_score: number;
  negative_score: number;
  neutral_score: number;
  processed_at: string;
  article_title?: string;
  article_source?: string;
  article_source_url?: string;
  article_event_category?: string;
}

export interface SentimentSummary {
  ticker: string | null;
  sector: string | null;
  total_articles: number;
  positive_count: number;
  negative_count: number;
  neutral_count: number;
  avg_positive: number;
  avg_negative: number;
  avg_neutral: number;
  dominant_label: string;
}

export interface SentimentTimePoint {
  date: string;
  avg_positive: number;
  avg_negative: number;
  avg_neutral: number;
  article_count: number;
  dominant_label: string;
}
